core/tee/tee_fs_rpc.c

```diff
@@ -0,0 +1,572 @@
+/*
+ * Copyright (c) 2016, Linaro Limited
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <assert.h>
+#include <kernel/thread.h>
+#include <mm/core_memprot.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tee/tee_fs.h>
+#include <tee/tee_fs_rpc.h>
+#include <trace.h>
+#include <util.h>
+
+struct tee_fs_dir {
+	int nw_dir;
+	struct tee_fs_dirent d;
+};
+
+static int tee_fs_rpc_send_cmd(int cmd_id, struct tee_fs_rpc *bf_cmd,
+			       void *data, uint32_t len, uint32_t mode)
+{
+	TEE_Result ret;
+	struct optee_msg_param params;
+	paddr_t phpayload = 0;
+	uint64_t cpayload = 0;
+	struct tee_fs_rpc *bf;
+	int res = -1;
```

Comment:
Yes, this function should probably return TEE_Result. I'll fix it.

Regarding the other functions, they simply return the status value set by tee-supplicant, unless communication failed in which case they return -1. The convention is: `<0` means failure (in this case the actual value is system-specific, can be -(errno) as explained in [1]), `>=0` is success and depending on the function it may be a useful value like a file descriptor or a number of bytes.

So... I propose to `#define RPC_FAILED -1` and use that instead of the hardcoded  `-1` to initalize `int rc = -1;` in the below functions (`tee_fs_rpc_read` and so on). And, to add the above explanation as a comment in `tee_fs_rpc.h`. Does it sound OK?

[1] https://github.com/OP-TEE/optee_client/pull/50/files#diff-4b1fc55b89725a5c4b5b364946cc554eR412